src/sys/system_v/shm.rs

```diff
@@ -0,0 +1,448 @@
+//! Safe wrapper around a SystemV shared memory segment
+//!
+
+use std::{
+    mem::ManuallyDrop,
+    ops::{Deref, DerefMut},
+    ptr::{null, null_mut},
+};
+
+use crate::Result;
+use crate::{errno::Errno, sys::stat::Mode};
+
+use libc::{self, c_int, c_void, key_t, shmid_ds};
+
+#[derive(Debug)]
+/// Safe wrapper around a SystemV shared memory segment
+///
+/// The shared memory segment size is equal to the size of T.
+///
+/// This is a smart pointer, and so implement the [`Deref`] and [`DerefMut`] traits.
+/// This means that you can work with the shared memory zone like you would with a [`Box`].
+///
+/// This type does not automatically create or destroy a shared memory segment,
+/// but only attach and detach from them using RAII.
+///
+/// To create one, use [`SharedMemory::shmget`], with the key [`ShmgetFlag::IPC_CREAT`].\
+/// To delete one, use [`SharedMemory::shmctl`], with the key [`ShmctlFlag::IPC_RMID`].
+///
+/// # Example
+///
+/// ```no_run
+/// # use nix::errno::Errno;
+/// # use nix::sys::system_v::shm::*;
+/// # use nix::sys::stat::Mode;
+/// #
+/// struct MyData(i64);
+/// const MY_KEY: i32 = 1337;
+///
+/// let id = SharedMemory::<MyData>::shmget(
+///     MY_KEY,
+///     ShmgetFlag::IPC_CREAT | ShmgetFlag::IPC_EXCL,
+///     Mode::S_IRWXU | Mode::S_IRWXG | Mode::S_IRWXO,
+/// )?;
+/// let mut shared_memory = SharedMemory::<MyData>::new(
+///     id,
+///     None,
+///     ShmatFlag::empty(),
+///     Mode::empty(),
+/// )?;
+///
+/// // This is writing to the stored [`MyData`] struct
+/// shared_memory.0 = 0xDEADBEEF;
+/// # Ok::<(), Errno>(())
+/// ```
+///
+pub struct SharedMemory<T> {
+    id: i32,
+    shm: ManuallyDrop<Box<T>>,
+}
+
+impl<T> Deref for SharedMemory<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.shm
+    }
+}
+impl<T> DerefMut for SharedMemory<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.shm
+    }
+}
+
+impl<T> Drop for SharedMemory<T> {
+    fn drop(&mut self) {
+        Self::shmdt(self).expect("SharedMemory detach from SysV IPC");
+    }
+}
+
+impl<T> SharedMemory<T> {
+    /// Create a new SharedMemory object
```

Comment:
This should probably removed considering that it won't create a new segment but attach to an existing one.

Reply:
Fixed by removing the ctor and letting `Shm` wraps around it.